contrib/slurm/README.md

```diff
@@ -0,0 +1,86 @@
+
+## Usage
+1. Run the script with Python 3:
+   ```
+   python3 $BENCHEXEC_FOLDER/contrib/slurm-benchmark.py [options]
+   ```
+   Options:
+   - `--slurm`: Use SLURM to execute benchmarks. Will revert to regular (local) benchexec if not given. 
+   - `--singularity <path_to_sif>`: Specify the path to the Singularity .sif file to use. See usage later.
+   - `--scratchdir <path>`: Specify the directory for temporary files. The script will use this parameter to create temporary directories for file storage per-run, which get discarded later. By default, this is the CWD, which might result in temporary files being generated by the thousands in the working directory. On some systems, this must be on the same mount, or even under the same hierarchy as the current directory. Must exist, be writable, and be a directory.
+   - `-N <N>`: Specify the factor of parallelism, i.e., how many instances to start at a time. Tested with up to `1000`, probably works with much higher values as well.
+    
+## Overview of the Workflow
+
+The workflow is based on `localexecution.py`, and uses the same general layout. However, instead of delegating to `runexec`, it delegates to `srun`.
+
+1. If the `--singularity` option is given, the script wraps the command to run in a container. This is useful for dependency management (in most HPC environments, arbitrary package installations are frowned upon). For a simple container, use the following: 
+
+    ```singularity
+    BootStrap: docker
+    From: ubuntu:22.04
+    
+    %post
+    apt -y update
+    apt -y install openjdk-17-jre-headless libgomp1 libmpfr-dev fuse-overlayfs
+    ```
+   
+    Use `singularity build [--remote / --fakeroot] --fix-perms <name>.sif <name>.def` to build the container.
+    
+    Notice the `fuse-overlayfs` package. That is mandatory for the overlay filesystem to work properly.
+    
+    The script parameterizes `singularity exec` with the following params:
+    * `-B $PWD:/lower`: Bind the working directory to `/lower` (could be read-only)
+    * `--no-home`: Do not bind the home directory
+    * `-B {tempdir}:/overlay`: Bind the temporary directory to `/overlay` (must be writeable)
+    * `--fusemount  "container:fuse-overlayfs -o lowerdir=/lower -o upperdir=/overlay/upper -o workdir=/overlay/work $HOME"`: mount an overlay filesystem at $HOME, where modifications go in the temp dir but files can be read from the current dir
+    
+    Modifications to the script are necessary if the user on the host and the container differ (due to using $HOME). Also, files not under the current directory will not be visible.
+
+2. Currently, the following parameters are passed to `srun` (calculated from the benchmark's parameters):
+   * `-t <hh:mm:ss>` CPU timelimit (generally, SLURM will round up to nearest minute)
+   * `-c <cpus>` number of cpus 
+   * `--threads-per-core=1` only use one thread per core
```

Comment:
BenchExec typically does not enforce this, only if you pass `--no-hyperthreading`. Can we have the same behavior here? The more that `slurm-benchmark.py` is a drop-in replacement for `benchexec` with the same behavior, the better.

Reply:
I'm unsure whether that is a good idea, because unfortunately, I cannot query the nodes about the number of threads per core -- we _can_ assume that it's always 2, but that is not a good idea.
The problem with not knowing this number is that I have to request whole cores from SLURM, and conventionally, benchexec requests threads (correct me if I'm mistaken!). Therefore, I have to divide the requested number of threads by the threads-per-core number to get the desired amount of CPU cores.
Another option is to pass `--threads-per-core=2` and `-c <cpus/2>` by default, and `--threads-per-core=1` and `-c <cpus>` when the `--no-hyperthreading` flag is given. However, this may confuse people who work with non-hyperthreading CPUs in their environment, as their jobs will simply refuse to start, as no nodes will match the desired resource allocation.

Reply:
I understand. Indeed, in BenchExec "cores" means virtual cores (threads in SLURM terms).  And we should not assume that the number of virtual cores per physical core is always 2.

So how about we require users to explicitly pass `--no-hyperthreading` and keep the current `--threads-per-core=1`? If users do not pass `--no-hyperthreading`, we can print an error message telling them to do so and exit. This avoids silently running benchmarks in a different configuration than the users expect, at the cost of a minor inconvenience. But it would also allow us to improve the situation without compatibility problems if we find a solution in the future.

A note in the limitations would also be good.

Reply:
That's a good idea, I'll do this.